### PR TITLE
Add fancy user interface

### PR DESCRIPTION
**Pull request by RauliL**
Introduces more fancier user interface for the game engine that uses CSS3 animations. Assets are compiled with Webpack.

### CODE REVIEW

File: ui/fancy/src/js/index.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+  function update(state) {
+    state.blocks.forEach((block, index) => {
+      const element = blockElements[index];
+
+      element.classList.remove('red', 'green', 'blue');
+      if (block.color) {
+        element.classList.add(block.color);
+      }
+      if (block.swapTimer === -2 || block.swapTimer === 2) {
```
Comment (frostburn) [inline ui/fancy/src/js/index.js]: You can't really depend on the swap timer having specific break points like this as it depends on the used ruleset.

**Comment by RauliL** [comment]
This one is ready for review.

File: ui/fancy/src/css/style.less [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+@color-background-grid: #2d0238;
+@color-block-red: #a7383e;
+@color-block-green: #2c8437;
+@color-block-blue: #303e73;
+@color-swapper: #ffffff;
+
+@grid-width: 6;
+@grid-height: 8;
+
+@block-size: 40px;
```
Comment (frostburn) [inline ui/fancy/src/css/style.less]: I think we should measure things in ems as we are populating the blocks with text content.

**Comment by frostburn** [comment]
A readme would be nice.

**Comment by frostburn** [comment]
A preview for the next row is needed for playability. A chain number counter would also be nice.

**Comment by frostburn** [comment]
Support for a slower flowing ruleset would also be nice, but is not required at the moment.

**Comment by frostburn** [comment]
When adding a row the swapper should go up one unit to stay in the same place relative to the blocks as they rise.

**Comment by frostburn** [comment]
The gamepad interface should only register the initial button presses. Now it goes into a crazy swapping frenzy when I press the button on my gamepad on firefox.

File: ui/fancy/src/js/ui.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+          continue;
+        }
+        this.gamepadSupport.previousTimestamps[index] = timestamp;
+
+        for (let i = 0; i < device.buttons.length; ++i) {
+          if (!device.buttons[i].pressed) {
+            continue;
+          } else if (i === 0) {
+            this.action('swap');
+          } else if (i === 1) {
```
Comment (frostburn) [inline ui/fancy/src/js/ui.js]: The row adding buttons are traditionally the shoulder buttons so they could be added alongside.

Comment (RauliL) [inline ui/fancy/src/js/ui.js]: Index of shoulder buttons varies between different game pads so there is no portable way to do so unless some third party library containing database of different game pads is included in the user interface.

Comment (RauliL) [inline ui/fancy/src/js/ui.js]: `0` and `1` are X and A on my SNES controller so I used them instead.

**Comment by frostburn** [comment]
I guess the issue does not affect all gamepads so we'll just blame the experimental API.
I don't see any obstacles for merging now.

**Pull request merged**